activerecord/lib/active_record/relation.rb

```diff
@@ -294,15 +294,21 @@ def any?
     # Returns true if there is exactly one record.
     def one?
       return super if block_given?
-      limit_value ? records.one? : size == 1
+      limit_value ? records.one? : flat_size == 1
     end
 
     # Returns true if there is more than one record.
     def many?
       return super if block_given?
-      limit_value ? records.many? : size > 1
+      limit_value ? records.many? : flat_size > 1
     end
 
+    def flat_size
+      size_val = size
+      size_val.is_a?(Hash) ? size_val.length : size_val
```

Comment:
I don't think this handles correctly multiple items of the same group key.
You can reproduce the error by adding `assert_predicate posts.group(:tags_with_destroy_count), :many?` to `def test_many_with_group` and see that it incorrectly fails.

The correct way of dealing with `Hash` is to sum the values, for example:
```suggestion
      size_val.is_a?(Hash) ? size_val.values.sum : size_val
```

Reply:
Thanks for reviewing! Why would you expect `:many?` to return true for `posts.group(:tags_with_destroy_count)`?

Here's what I get for the test data in this case:

```ruby
  posts.length # is 11
  posts.group(:tags_with_destroy_count).length # is 1
  posts.group(:tags_with_destroy_count).count # is {0=>11}
```

All records have a `tags_with_destroy_count` value of zero.

Reply:
Isn't `many?` supposed to return `true` when having more than 1 record as result? For example here we have 11 records as result, even if they have the same `:tags_with_destroy_count` value.

Reply:
But `posts.group(:tags_with_destroy_count).length` returns 1 and `.length` uses the actual records returned.

GROUP BY, as I understand it, returns a relation with as many records as there are unique values in the column in question. If all records in your example have the value zero, the relation returned has a single row.

The behaviour of `.count` combined with `.group`, which is what produces the Hash, is very interesting in itself. For your example, `.count` returns `{0=>11}` where 0 is the unique value count groups by and 11 is the number of rows within this group.

Here are some extreme examples from another database:

```
[1] pry(main)> Provider.group(:code, :id).count
   (0.4ms)  SELECT COUNT(*) AS count_all, "providers"."code" AS providers_code, "providers"."id" AS providers_id FROM "providers" GROUP BY "providers"."code", "providers"."id"
=> {["SF2", 22]=>1,
 ["2A0", 15]=>1,
 ["4AB", 19]=>1,
 ["S72", 5]=>1,
 ["D39", 4]=>1,
 ["1N1", 10]=>1,
 ["1JB", 6]=>1,
 ["4HP", 14]=>1,
 ["20C", 13]=>1,
 ["24P", 2]=>1,
 ["W53", 7]=>1,
 ["IMZ", 20]=>1,
 ["1JA", 1]=>1,
 ["24W", 18]=>1,
 ["SDM", 27]=>1,
 ["M40", 8]=>1,
 ["TC3", 11]=>1,
 ["4T7", 9]=>1,
 ["YUJ", 26]=>1,
 ["X8E", 21]=>1,
 ["24J", 3]=>1,
 ["1UY", 17]=>1,
 ["BZ5", 16]=>1,
 ["AN0", 12]=>1,
 ["Q66", 24]=>1,
 ["HQG", 25]=>1,
 ["0C3", 23]=>1}
```

and

```
pry(main)> Provider.joins(:courses).group(:code, :id).count
   (0.5ms)  SELECT COUNT(*) AS count_all, "providers"."code" AS providers_code, "providers"."id" AS providers_id FROM "providers" INNER JOIN "courses" ON "courses"."provider_id" = "providers"."id" GROUP BY "providers"."code", "providers"."id"
=> {["S72", 5]=>36,
 ["D39", 4]=>10,
 ["1N1", 10]=>97,
 ["1JB", 6]=>41,
 ["24P", 2]=>14,
 ["1JA", 1]=>226,
 ["4T7", 9]=>9,
 ["24J", 3]=>15}
```

It doesn't matter what the keys used for grouping look like, the number of records returned corresponds to the number of keys.

Am I missing something? It's a little confusing at first but I think it makes sense.

Reply:
I understand your point, it just looks odd to me that `any?/many?` behaves differently when `.group` is used.

Reply:
Thanks, the important thing is deciding how these methods are meant to behave when `.group` is part of the query. I think what I've suggested is reasonable but let's see what other people think, too.

As a aside, maybe `flat_size` as a name for the private method is confusing, because with a name like this you'd probably expect it to perform `size_val.values.sum`. Maybe something like `group_safe_size` would be more descriptive.

Reply:
> As a aside, maybe `flat_size` as a name for the private method is confusing, because with a name like this you'd probably expect it to perform `size_val.values.sum`. Maybe something like `group_safe_size` would be more descriptive.

How about `number_of_rows`?

Reply:
Thanks for the suggestion. I've gone for `number_of_rows_from_size`, just to make it clear `#size` is involved, as there many different ways we establish the number of rows in this code, for example:

```ruby
limit_value ? records.many? : number_of_rows_from_size > 1
```